### PR TITLE
flexible grid to streamline affine generation and pathlength function

### PR DESCRIPTION
**Pull request by kesshijordan**
Included are two related additions :

NF: The path_length function computes the shortest path, along any streamline, between the roi and each voxel. @MrBago and I are using this for a particular project we're working on, but it could be useful in other applications.

NF: The other supporting functions (flexi tvis affine, etc.) are used in path_length, but also address a problem that I've found particularly difficult as a going-up-the-learning-curve user: mapping between grid and streamline spaces with different voxel orders. After the change to using affines instead of providing voxel sizes to functions like target, one has to sort through a lot of information in the headers to get everything in the same space. These flexi tvis affine functions are extremely helpful for painless targeting and fvtk renderings.

### CODE REVIEW

File: dipy/tracking/utils.py [hunk 1]
```diff
@@ -956,3 +958,124 @@ def reduce_rois(rois, include):
             exclude_roi |= rois[i]
 
     return include_roi, exclude_roi
+
+def flexi_tvis_affine(sl_vox_order, grid_affine, dim, voxel_size):
+    """ Computes the mapping from voxel indices to streamline points,
+        reconciling streamlines and grids with different voxel orders
```
Comment (Garyfallidis) [inline dipy/tracking/utils.py]: indent to the left

File: dipy/tracking/utils.py [hunk 2]
```diff
@@ -956,3 +958,124 @@ def reduce_rois(rois, include):
...
+
+    Parameters
+    ----------
+    sl_vox_order: string of length 3
```
Comment (Garyfallidis) [inline dipy/tracking/utils.py]: add space after parameter and before ``:``
please change everywhere 
it should be ```sl_vox_order : string```

File: dipy/tracking/utils.py [hunk 3]
```diff
@@ -956,3 +958,124 @@ def reduce_rois(rois, include):
...
+        a string that describes the voxel order of the streamlines (ex: LPS)
+    grid_affine: nii_aff: array (4, 4),
```
Comment (Garyfallidis) [inline dipy/tracking/utils.py]: two parameters?

File: dipy/tracking/utils.py [hunk 4]
```diff
@@ -956,3 +958,124 @@ def reduce_rois(rois, include):
...
+        An affine matrix describing the current space of the grid in relation to RAS+ scanner space
+    dim: tuple of length 3
```
Comment (Garyfallidis) [inline dipy/tracking/utils.py]: add space (as previously suggested)

File: dipy/tracking/utils.py [hunk 5]
```diff
@@ -956,3 +958,124 @@ def reduce_rois(rois, include):
...
+        dimension of the grid
+    voxel_size: array (3,0)
```
Comment (Garyfallidis) [inline dipy/tracking/utils.py]: same

File: dipy/tracking/utils.py [hunk 6]
```diff
@@ -956,3 +958,124 @@ def reduce_rois(rois, include):
...
+        voxel size of the grid
+
+    Returns
+    -------
+    flexi_tvis_aff: this affine maps between a grid and a trackvis space
```
Comment (Garyfallidis) [inline dipy/tracking/utils.py]: same same

File: dipy/tracking/utils.py [hunk 7]
```diff
@@ -956,3 +958,124 @@ def reduce_rois(rois, include):
...
+
+    return flexi_tvis_aff
+
+def get_flexi_tvis_affine(tvis_hdr, nii_aff):
+    """ Computes the mapping from voxel indices to streamline points,
+        reconciling streamlines and grids with different voxel orders
+
+    Parameters
+    ----------
+    tvis_hdr: header from a trackvis file
```
Comment (Garyfallidis) [inline dipy/tracking/utils.py]: same

Comment (Garyfallidis) [inline dipy/tracking/utils.py]: same... (change everywhere in your files)

**Comment by Garyfallidis** [comment]
Also you may want to look at the new streamline API available in DIPY. This should solve many different problems with dealing with streamline datasets.

**Comment by arokem** [comment]
On Sat, Sep 10, 2016 at 9:43 PM, Eleftherios Garyfallidis <
notifications@github.com> wrote:

> Also you may want to look at the new streamline API available in DIPY.
> This should solve many different problems with dealing with streamline
> datasets.
>
>
Do you mean the new streamline API in *nibabel*? Worth taking a look at
that, for sure:
http://nipy.org/nibabel/reference/nibabel.streamlines.html#module-nibabel.streamlines

File: dipy/tracking/utils.py (hunk 3 above)
Comment (kesshijordan) [inline dipy/tracking/utils.py]: I'm interpreting this as "why do you include both the sl_vox_order and the grid_affine as parameters?" The streamline voxel order doesn't necessarily match that of the affine of the underlying grid one is trying to target/display in fvtk... the function called here (get_flexi_tvis_affine) converts both to orientations, then reconciles them with an affine.  The orientations can be different, so must be pulled from both the nifti and the trk headers.

File: dipy/tracking/utils.py (hunk 2 above)
Comment (kesshijordan) [inline dipy/tracking/utils.py]: fixed

File: dipy/tracking/utils.py (hunk 4 above)
Comment (kesshijordan) [inline dipy/tracking/utils.py]: fixed

File: dipy/tracking/utils.py (hunk 5 above)
Comment (kesshijordan) [inline dipy/tracking/utils.py]: fixed

File: dipy/tracking/utils.py (hunk 6 above)
Comment (kesshijordan) [inline dipy/tracking/utils.py]: fixed fixed

File: dipy/tracking/utils.py (hunk 7 above)
Comment (kesshijordan) [inline dipy/tracking/utils.py]: fixed

File: dipy/tracking/utils.py (hunk 1 above)
Comment (kesshijordan) [inline dipy/tracking/utils.py]: What do you mean? I think the def is against the left edge, and the docstring is indented like the rest of the functions in this file?

**Comment by kesshijordan** [comment]
Thank you for the comments. I have fixed most of the formatting issues (not sure I know what you mean regarding indentation), and commented on the number of parameters question. I also noticed that there are supposed to be two lines between each function, so I corrected that, as well. I will check out the nibabel API to see if that fixes some of these issues.

**Comment by arokem** [comment]
Hey @kesshijordan : do you understand why these test failures are happening?

**Comment by kesshijordan** [comment]
Hi, @arokem. Thanks for checking in. I think I figured it out: I found errors I recognized when I scrolled to the bottom of the Travis build outputs. I addressed them, resubmitted to Travis, and only one build failed (#1.5; I think because of the numpy version) "AttributeError: 'numpy.ufunc' object has no attribute 'at'", referring to a use of np.minimum.at(a,(i,j,k),b)... so I replaced it with a[i,j,k]=np.minimum(a[i,j,k],b) so that all of the builds work. Is there anything else I should be looking for besides the big ERRORs?

**Comment by kesshijordan** [comment]
@MrBago just pointed out that substituting a[i,j,k]=np.minimum(a[i,j,k],b) for np.minimum.at(a,(i,j,k),b) introduces errors and recommended fixing it with a little function that replaces np.minimum.at if the numpy version is less than 8.

File: dipy/tracking/utils.py [hunk 8]
```diff
@@ -956,3 +958,149 @@ def reduce_rois(rois, include):
...
+def flexi_tvis_affine(sl_vox_order, grid_affine, dim, voxel_size):
+    """ Computes the mapping from voxel indices to streamline points,
+        reconciling streamlines and grids with different voxel orders
+
+    Parameters
+    ----------
+    sl_vox_order : string of length 3
+        a string that describes the voxel order of the streamlines (ex: LPS)
+    grid_affine : array (4, 4),
+        An affine matrix describing the current space of the grid in relation to RAS+ scanner space
```
Comment (arokem) [inline dipy/tracking/utils.py]: PEP8

Comment (arokem) [inline dipy/tracking/utils.py]: Line's too long, I believe.

File: dipy/tracking/utils.py [hunk 9]
```diff
@@ -956,3 +958,149 @@ def reduce_rois(rois, include):
...
+        voxel size of the grid
+
+    Returns
+    -------
+    flexi_tvis_aff : this affine maps between a grid and a trackvis space
+    """
+
+    sl_ornt = orientation_from_string(str(sl_vox_order))
+    grid_ornt = nib.io_orientation(grid_affine)
+    reorder_grid = reorder_voxels_affine(grid_ornt, sl_ornt, np.array(dim)-1, np.array([1,1,1]))
```
Comment (arokem) [inline dipy/tracking/utils.py]: Here as well: line is longer than 80 characters (PEP8)

File: dipy/tracking/utils.py [hunk 10]
```diff
@@ -956,3 +958,149 @@ def reduce_rois(rois, include):
...
+
+def get_flexi_tvis_affine(tvis_hdr, nii_aff):
+    """ Computes the mapping from voxel indices to streamline points,
+        reconciling streamlines and grids with different voxel orders
+
+    Parameters
+    ----------
+    tvis_hdr : header from a trackvis file
+    nii_aff : array (4, 4),
+        An affine matrix describing the current space of the grid in relation to RAS+ scanner space
```
Comment (arokem) [inline dipy/tracking/utils.py]: PEP8 (line length)

File: dipy/tracking/utils.py [hunk 11]
```diff
@@ -956,3 +958,149 @@ def reduce_rois(rois, include):
...
+    nii_data : nd array
+        3D array, each with shape (x, y, z) corresponding to the shape of the brain volume,
```
Comment (arokem) [inline dipy/tracking/utils.py]: Period instead of comma at the end of the sentence?

File: dipy/tracking/tests/test_utils.py [hunk 12]
```diff
@@ -608,3 +608,98 @@ def test_reduce_rois():
...
+    [2.43251352e-03,   1.09376717e+00,   1.48301506e-02,  -1.07367630e+02],
+    [1.33170187e-01,  -8.34854878e-03,   1.98454463e+00,  -9.98151169e+01],
+    [0.00000000e+00,   0.00000000e+00,   0.00000000e+00,   1.00000000e+00]])
+    dim = (256,256,86)
+    voxel_size = np.array([ 1.09379995,1.09379995,1.99947774])
+    affine = flexi_tvis_affine(sl_vox_order, grid_affine, dim, voxel_size)
+
+    origin = np.dot(affine, [0, 0, 0, 1])
+    assert_array_almost_equal(origin[:3], np.multiply(dim,voxel_size)-voxel_size/2)
+
```
Comment (arokem) [inline dipy/tracking/tests/test_utils.py]: One more linebreak here (PEP8)

File: dipy/tracking/tests/test_utils.py [hunk 13]
```diff
@@ -608,3 +608,98 @@ def test_reduce_rois():
...
+    #grid_affine =
+    tvis_hdr['voxel_order'] = 'ASL'
+    vsz = tvis_hdr['voxel_size'] = np.array([3, 4, 2.])
+    affine = get_flexi_tvis_affine(tvis_hdr, grid_affine)
+
+    vox_point = np.array([9, 8, 7])
+    trk_point = np.dot(affine, np.append(vox_point, 1))
+
+    assert_array_almost_equal(trk_point[:3],  (vox_point[[1, 2, 0]] + 0.5) * vsz)
+
```
Comment (arokem) [inline dipy/tracking/tests/test_utils.py]: One more linebreak here (PEP8)

**Comment by arokem** [comment]
Thanks! Please do consider adding documentation examples using these functions, so that we can see how exactly this is useful to you. If you need a hand with that, just let me know.

**Comment by kesshijordan** [comment]
Sorry this took so long, @arokem. I just made a PR for the documentation that goes with this #1631.

**Pull request merged**